### ISSUE TITLE
v0.27 facetsDistribution behavior changed on empty response

### ISSUE THREAD

**Issue by bidoubiwa** [meilisearch/meilisearch#2352]
**Describe the bug**
In v0.26 when a search was made with the facetsDistribution set to `genres` for example, even when the returned hits contained no hits contained `genres` the facets distribution value was: 

```js
    facetsDistribution: { genres: {} },
```

with the v0.27 it becomes: 

```js
    facetsDistribution: {},
```

Works in:
- v0.26.1

but not in: 
- V0.27.0rc0
- V0.27.0rc1
- V0.27.0rc2

**To Reproduce**
1. add `genres` in the `filterableAttributes` of an empty index.
2. add one document lets say: `{ "id": 1, "title": "hello" }`
3. make a search with `facetDistribution` set to `["genres"]`
4. this is the return `facetsDistribution: {}`

**Expected behavior**
Same behavior as before

**Meilisearch version:** [e.g. v0.20.0]
v0.27.0rc2

**Comment by curquiza** [comment]
@gmourier is this behavior in the specs? A thing like this "all the required attributes in `facetsDistribution` (request body) should be put in the search response body in `facetDistribution`"
This behavior is important for front-end developer if I'm correct, can you confirm @bidoubiwa?

Also can we add tests about this behavior? 🔥

**Comment by irevoire** [comment]
```json
{
  "id": "the best doggo",
  "doggo": {
    "name": "bob",
    "race": "bernese mountain",
    "picture": "https://thewoofclub.ch/wp-content/uploads/2020/09/bernese-puppy.png"
  }
}
```

If I ask the field distribution for `doggo`. What should happen?
```json
{
  "doggo": {},
  "doggo.name": 1,
  "doggo.race": 1,
  "doggo.picture": 1,
}
```

Or
```json
{
  "doggo.name": 1,
  "doggo.race": 1,
  "doggo.picture": 1,
}
```
?

@gmourier @bidoubiwa

**Comment by gmourier** [comment]
For me it is solution 1 that should be implemented in this case because you could have a non-nested `doggo` field in another document.

This will happen in case a user has set `doggo` as a filterable and not just a specific field like `doggo.name`.

**Comment by bidoubiwa** [comment]
I m not sure I understand the context. Was `doggo` added as a `filterableAttribute`?  How does that work if I want to filter on it? 
`Doggo = "foe"`. If I want to have the facets distribution of `doggo.name` do I have to add it to  `facetsDistribution: ["doggo.name"]` or does `facetsDistribution: ["doggo"]` makes all child filterable? If yes to which extend ? The nested of the nested of the nested if filterable?

I'm a bit lost.

**Comment by gmourier** [comment]
> Was `doggo` added as a filterableAttribute?

Yep otherwise it would have throw an error at search.

> filter = Doggo = "foe" what happens in this case ?

In this case, it will return documents having a simple `doggo` field containing `foe`

> If I want to have the facets distribution of `doggo.name` do I have to add it to `facetsDistribution: ["doggo.name"]` or does `facetsDistribution: ["doggo"]` makes all child filterable?

This is the last one, `facetsDistribution: ["doggo"]` permits to ask for a distribution on `doggo.name` if the field exists as well has `doggo` to distribute facets from all nested fields.

We can expect that a large majority of users will not mix documents containing a single field `doggo` and others with `doggo` as a nested object.

In any case, the engine just tries to adapt as best as possible to the fact that a schema does not exist. If you specify `doggo.name` instead of `doggo` as a filterableAttributes, the engine will throw an error if `facetsDistribution` contains `doggo`. This is the current behavior for a field that has not been set as a filterableAttributes.

**Comment by gmourier** [comment]
> Was `doggo` added as a filterableAttribute?

Yep otherwise it would have thrown an error at search.

> If I want to have the facets distribution of `doggo.name` do I have to add it to `facetsDistribution: ["doggo.name"]` or does `facetsDistribution: ["doggo"]` makes all child filterable?

This is the last one, `"filterableAttributes": ["doggo"]` permits to ask for a distribution on `doggo.name` if the field exists in `doggo` as well than asking for a distribution for all nested fields by using `"facetsDistribution": ["doggo"]`.

We can expect that a majority of users will not mix documents containing a single field `doggo` and others with `doggo` as a nested object.

I**n any case, the engine just tries to adapt as best as possible to the fact that a schema does not exist.** If you specify `doggo.name` instead of `doggo` as a filterableAttributes, the engine will throw an error if `filter` and `facetsDistribution` contains `doggo` instead of `doggo.name`. This is the current behavior for a field that has not been set as a `filterableAttributes`.

**Comment by gmourier** [comment]
Testing with v0.26.1 I have the current v0.27RC behavior with postman.

![Capture d’écran 2022-04-27 à 16 37 13](https://user-images.githubusercontent.com/3692335/165544000-bc70a477-a7bb-4959-b0a9-085f3ea9d091.png)

Documents do not contain a `genre` field. Did you tested it doing a pure API call?

**Comment by bidoubiwa** [comment]
This is problematic as instant-search uses facetsDistribution to create components and excepts at all time a `key: { ... }`

Another inconsistency is that the following implies that `name` is a facet value. But a `key` is not a facet value.
```
{
  "doggo": {},
  "doggo.name": 1,
  "doggo.race": 1,
  "doggo.picture": 1,
}
```


that output makes the most sense for me: 

```
doggo: {}
doggo.name: {},
doggo.race: {},
doggo.picture: {}
```

and if you have results: 


```
doggo: {}
doggo.name: { "medor" : 1 },
doggo.race: {},
doggo.picture: {}
```

**Comment by bidoubiwa** [comment]
You can reproduce it the following way:

- create a repo
- `npm init -y`
- `npm install meilisearch`
- `touch index.js`
- open index.js copy the following:
```js
const { MeiliSearch } = require('meilisearch')

;(async () => {
  const client = new MeiliSearch({
    host: 'http://127.0.0.1:7700',
    apiKey: 'masterKey',
  })

  // An index is where the documents are stored.
  const index = client.index('movies')

  const dataset = [
      { id: 1, title: 'Carol' },
      { id: 2, title: 'Wonder Woman', genres: ['Action', 'Adventure'] },
      { id: 3, title: 'Life of Pi', genres: ['Adventure', 'Drama'] },
      { id: 4, title: 'Mad Max: Fury Road', genres: ['Adventure', 'Science Fiction'] },
      { id: 5, title: 'Moana', genres: ['Fantasy', 'Action']},
      { id: 6, title: 'Philadelphia', genres: ['Drama'] },
  ]

  // If the index 'movies' does not exist, MeiliSearch creates it when you first add the documents.
  await index.updateFilterableAttributes([
    'director',
    'genres'
  ])

  let { uid } = await index.addDocuments(dataset)

  await client.waitForTask(uid)

  const filteredSearch = await index.search('Carol', {
    // attributesToHighlight: ['*'],
    facetsDistribution: ["genres"]
  })
  console.log({ filteredSearch, hit: filteredSearch.hits[0] })

})()
```
- run `node index.js` 


Try with v0.26 and v0.27

**Comment by gmourier** [comment]
In this example, your documents contain a `genres` attribute. Okay I will investigate. Thanks

**Comment by bidoubiwa** [comment]
I made a search on `Carol` on which there is not genres.

**Comment by gmourier** [comment]
I managed to reproduce it.

Here is the behavior of v0.26.0:

- If no documents in the db have a `genres` field, asking for `genres` as a facetDistribution won't add it in the response.
- If some documents in the db have a `genres` field but the returned search results won't have a `genres` field. `genres` is added as an empty object in `facetsDistribution`.

**Comment by gmourier** [comment]
If we have to talk about the behavior of nested fields with facetsDistribution, please make another issue if the current behavior is not good after the bug has been fixed for the sake of clarity. We wasted a lot of time responding and trying to figure out the bug which was not related to the nested field support feature initially. 😇

**Comment by curquiza** [comment]
Closed by #2366
